subworkflows/nf-neuro/io_bids/main.nf

```diff
@@ -0,0 +1,88 @@
+include { IO_READBIDS   } from '../../../modules/nf-neuro/io/readbids/main.nf'
+
+workflow IO_BIDS {
+
+    take:
+    bids_folder         // channel: [ path(bids_folder) ]
+    fs_folder           // channel: [ path(fs_folder) ]
+    bidsignore         // channel: [ path(bids_ignore) ]
+
+    main:
+
+    ch_versions = Channel.empty()
+
+    // ** Fetching the BIDS data as a json file ** //
+    IO_READBIDS (
+        bids_folder,
+        fs_folder,
+        bidsignore
+    )
+    ch_versions = ch_versions.mix(IO_READBIDS.out.versions)
+
+    // ** Converting the json file into channels. ** //
+    ch_files = IO_READBIDS.out.bidsstructure
+        .flatMap{ layout ->
+            def json = new groovy.json.JsonSlurper().parseText(layout.getText())
+            json.collect { item ->
+                // ** Collecting the subject's ID ** //
+                def sid = "sub-" + item.subject
+
+                // ** Collecting the session's ID if present ** //
+                def ses = item.session ? "_ses-" + item.session : ""
+
+                // ** Collecting the run's ID if present ** //
+                def run = item.run ? "run-" + item.run : ""
+
+                // ** Collecting TotalReadoutTime, PhaseEncodingDirection ** //
+                def tr = item.TotalReadoutTime ?: ""
+                def phase = item.DWIPhaseEncodingDirection ?: ""
+                def revphase = item.rev_DWIPhaseEncodingDirection ?: ""
+
+                // ** Validating there is not missing data ** //
+                item.each { _key, value ->
+                    if (value == "todo") {
+                        error   "ERROR ~ $sid contains missing files, please" +
+                                "check the BIDS layout. You can validate your" +
+                                "structure using the bids-validator tool:" +
+                                "https://bids-standard.github.io/bids-validator/" +
+                                "or the docker version: " +
+                                "https://hub.docker.com/r/bids/validator"
+                    }
+                }
+
+                // ** Collecting the files ** //
+                return [
+                    [id: sid, session: ses, run: run, tr: tr, phase: phase, revphase: revphase],
```

Comment:
I'd like it, but I think having parameters like TR that vary between modalities will break channel jointing.

Reply:
I would be fine with adding an extra map to the outputs of IO_BIDS. There isn't a way to process it for us, but for now we can garbage it and start updating the modules that could use it.